pulsar/consumer.go

```diff
@@ -307,6 +307,14 @@ type Consumer interface {
 	// AckID the consumption of a single message, identified by its MessageID
 	AckID(MessageID) error
 
+	// AckIDList the consumption of a list of messages, identified by their MessageIDs
+	// Returns a map of MessageID to error, the keys are the MessageIDs that failed to be acknowledged
+	// NOTE: When EnableBatchIndexAcknowledgment is false, if a message ID represents a message in the batch,
+	// it will not be actually acknowledged by broker until all messages in that batch are acknowledged via
+	// the AckID or AckIDList method.
+	// However, in this case, no error will be returned for that message ID even if AckWithResponse is true.
+	AckIDList([]MessageID) map[MessageID]error
```

Comment:
Can we return the `error` directly to simplify this API?
And we can also define the error type like this to let users get more specified errors for each msg ids:
```go
type AckIDListError map[MessageID]error

func (e AckIDListError) Error() string {
    return "some message IDs failed to be acknowledged"
}
```

This way, the user can process the error like this:
```go
if err != nil {
        if ackErr, ok := err.(AckIDListError); ok {
	        for id, e := range ackErr {
		        fmt.Printf("Message ID %d: %s\n", id, e)
	        }
        }
}
```

Reply:
Users should not ignore any failed message IDs, that's why I return a map error. I don't want to users to call it like:

```golang
if err := consumer.AckIDList(ids); err != nil {
        fmt.Println(err)
}
```

Reply:
If it returns `error` directly, in which case should users not convert it to the `AckIDListError`?

Reply:
> Users should not ignore any failed message IDs, that's why I return a map error. I don't want users to call it like:

Yes. Users could handle it. And users should not call it like that.

Instead, I gave the example in my above comment:
```
if err != nil {
        if ackErr, ok := err.(AckIDListError); ok {
	        for id, e := range ackErr {
		        fmt.Printf("Message ID %d: %s\n", id, e)
	        }
        }
}
```
The users could check the failed messages ids my converting the error to `AckIDListError`.

> If it returns error directly, in which case should users not convert it to the AckIDListError?

So, if the error is nil, then the user doesn't need to conver it.

Reply:
I think `map[MessageID]error` is better. Usually, I don't check the error type.

Reply:
@nodece I just had a quick sync with @RobertIndie 

With the current implementation, the only case that only partial of the message ID list failed is: there are some invalid message IDs.

Therefore, we can just return a trivial `error` to tell users that all these message ID failed.

```golang
if err := consumer.AckIDList(ids); err != nil {
    if errorMap, ok := err.(AckIDListError); ok {
        for msgID, err := range errorMap {
            // Add msgID to the message ID list to ack again
        }
    } else {
        // Add the whole ids to the message ID list to retry
    }
}
```

WDYT?

Reply:
What about returnning an error that are not specified to the message ID level like the `ErrConsumerClosed`? It doens't make sense to make it as the message ID error.
Actually, checking the type of error should be a common practice. We have discussed this before: https://github.com/apache/pulsar-client-go/issues/1142. It is better to expose the error type to users so they can easily identify the error.

Reply:
Or should we just fail the whole ids if there is any invalid message ID?

I designed the `map[MessageID]error` as the return value type because I want to handle the case that partial messages in the batch are acknowledged. But since now, they are treated as successful result, it might not be meaningful to return a map now.

Reply:
> Can we return the `error` directly to simplify this API? And we can also define the error type like this to let users get more specified errors for each msg ids:
> 
> ```go
> type AckIDListError map[MessageID]error
> 
> func (e AckIDListError) Error() string {
>     return "some message IDs failed to be acknowledged"
> }
> ```
> 
> This way, the user can process the error like this:
> 
> ```go
> if err != nil {
>         if ackErr, ok := err.(AckIDListError); ok {
> 	        for id, e := range ackErr {
> 		        fmt.Printf("Message ID %d: %s\n", id, e)
> 	        }
>         }
> }
> ```

I guess many users will habitually directly judge `err != nil` and then print out the err information.

Maybe this would be better(or somthine about more infomations):

```go
type AckIDListError map[MessageID]error

func (e AckIDListError) Error() string {
	ids := make([]MessageID, 0)
	for id, err := range e {
		if err != nil{
			ids = append(ids, id)
		}
	}
	return fmt.Sprintf("some message IDs %s failed to be acknowledged", ids)
}
```

Reply:
I adopted this solution eventually, the difference is that I included the error message abd grouped the message IDs with the same error message.